src/main/java/com/intuit/graphql/orchestrator/schema/transform/KeyTransformer.java

```diff
@@ -0,0 +1,44 @@
+package com.intuit.graphql.orchestrator.schema.transform;
+
+import com.intuit.graphql.graphQL.Argument;
+import com.intuit.graphql.graphQL.Directive;
+import com.intuit.graphql.graphQL.TypeDefinition;
+import com.intuit.graphql.orchestrator.keydirective.KeyDirectiveValidator;
+import com.intuit.graphql.orchestrator.xtext.XtextGraph;
+import graphql.VisibleForTesting;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.intuit.graphql.orchestrator.utils.XtextUtils.FEDERATION_KEY_DIRECTIVE;
+import static com.intuit.graphql.orchestrator.utils.XtextUtils.typeContainsDirective;
+
+/**
+ * This class is responsible for checking the merged graph for any key directives. For each field in key
+ *  directive, this class will validate the fields to the key directive by ensuring they exist
+ */
+public class KeyTransformer implements Transformer<XtextGraph, XtextGraph> {
+
+  @VisibleForTesting
+  KeyDirectiveValidator validator = new KeyDirectiveValidator();
+
+  @Override
+  public XtextGraph transform(final XtextGraph source) {
+    List<TypeDefinition> entities = source.getTypes().values().stream()
+            .filter(typeDefinition -> typeContainsDirective(typeDefinition, FEDERATION_KEY_DIRECTIVE))
```

Comment:
just want to make sure we tried the union Entity added by federation libraries

Reply:
Didn't use the union type to get the entities, because I would then have to do another stream to retrieve the type definition, so it can be validated